src/Components/Aspire.Azure.Npgsql/AspireAzureNpgsqlExtensions.cs

```diff
@@ -0,0 +1,157 @@
+                var token = credential.GetToken(s_databaseForPostgresSqlTokenRequestContext, default);
+
+                if (TryGetUsernameFromToken(token.Token, out var username))
+                {
+                    dataSourceBuilder.ConnectionStringBuilder.Username = username;
+                }
+                else
+                {
+                    throw new InvalidOperationException("Could not determine username from token claims");
+                }
+            }
+
+            if (string.IsNullOrEmpty(dataSourceBuilder.ConnectionStringBuilder.Password))
+            {
+                // The token is not cached since it it refreshed for each new physical connection, or when it has expired.
+
+                dataSourceBuilder.UsePasswordProvider(
+                    passwordProvider: _ => credential.GetToken(s_databaseForPostgresSqlTokenRequestContext, default).Token,
+                    passwordProviderAsync: async (_, ct) => (await credential.GetTokenAsync(s_databaseForPostgresSqlTokenRequestContext, default).ConfigureAwait(false)).Token
+                );
+            }
+
+            // Delay validating the ConnectionString until the DataSource is requested. This ensures an exception doesn't happen until a Logger is established.
+            ConnectionStringValidation.ValidateConnectionString(settings.ConnectionString, connectionName, DefaultConfigSectionName);
+
+            configureDataSourceBuilder?.Invoke(dataSourceBuilder);
+        }, serviceKey: serviceKey);
+    }
+
+    private static bool TryGetUsernameFromToken(string jwtToken, out string? username)
+    {
+        username = null;
+
+        // Split the token into its parts (Header, Payload, Signature)
+        var tokenParts = jwtToken.Split('.');
+        if (tokenParts.Length != 3)
+        {
+            return false;
+        }
+
+        // The payload is the second part, Base64Url encoded
+        var payload = tokenParts[1];
+
+        // Add padding if necessary
+        payload = AddBase64Padding(payload);
+
+        // Decode the payload from Base64Url
+        var decodedBytes = Convert.FromBase64String(payload);
+
+        // Parse the decoded payload as JSON
+        var reader = new Utf8JsonReader(decodedBytes);
+        var payloadJson = JsonElement.ParseValue(ref reader);
+
+        // Try to get the username from 'upn', 'preferred_username', or 'unique_name' claims
+        if (payloadJson.TryGetProperty("upn", out var upn))
+        {
+            username = upn.GetString();
+        }
+        else if (payloadJson.TryGetProperty("preferred_username", out var preferredUsername))
+        {
+            username = preferredUsername.GetString();
+        }
+        else if (payloadJson.TryGetProperty("unique_name", out var uniqueName))
+        {
+            username = uniqueName.GetString();
+        }
+
+        return username != null;
+    }
+
+    private static string AddBase64Padding(string base64) => (base64.Length % 4) switch
```

Comment:
I thought you mentioned this is unnecessary and can be removed?

Reply:
Not in the case of JWT tokens.